### PR TITLE
feat: upgrade dependencies, add persistence and other fixes

### LINKED ISSUE #17: add persistence using BadgerDB or cometdb
we need to use BadgerDB (or cometdb that wraps multiple dbs) to persist the state of the `TransactionQueue`, `BatchQueue`, `lastBatchHash`, and `seenBatches`, so that restarting the sequencer does not lose the in-memory data.

### PR DESCRIPTION
**Pull request by gupadhyaya**
Fixes #17  #16 
Also upgrades dependencies to latest: go-sequencing and go-da

## Summary by CodeRabbit

- **New Features**
	- Added a command-line flag to specify the database path for enhanced configuration.
	- Integrated persistent storage using BadgerDB for batch and transaction management, allowing recovery after crashes.

- **Bug Fixes**
	- Improved error handling and response structures in the sequencer's methods, including validation for rollup IDs.

- **Documentation**
	- Updated tests to reflect changes in method signatures and request/response structures, enhancing clarity and maintainability.

- **Chores**
	- Updated dependency requirements in the project configuration.

### CODE REVIEW

**Review by MSevey — COMMENTED** [review]
Does this branch need to be updated/rebased? Seems like a lot of other changes.

File: sequencing/sequencer.go [hunk 1]
```diff
@@ -137,19 +302,146 @@
...
+		// Set lastBatchHash in memory from BadgerDB
+		return item.Value(func(val []byte) error {
+			hash = val
+			return nil
+		})
+	})
+	// Set the in-memory lastBatchHash after successfully loading it from DB
+	c.lastBatchHash = hash
+	return err
+}
```
[the remark this answers is not in the document]
Comment (gupadhyaya) [reply sequencing/sequencer.go:394]: nil is ok

File: sequencing/sequencer.go [hunk 2]
```diff
@@ -137,19 +302,146 @@
...
+			item := it.Item()
+			key := item.Key()
+			// Add the batch hash to the seenBatches map (for fast in-memory lookups)
+			c.seenBatches[string(key)] = struct{}{}
+		}
+		return nil
+	})
+
+	return err
+}
```
[the remark this answers is not in the document]
Comment (gupadhyaya) [reply sequencing/sequencer.go:416]: this is useful check

**Comment by gupadhyaya** [comment]
> Does this branch need to be updated/rebased? Seems like a lot of other changes.

can you elaborate what you mean here?

File: sequencing/sequencer.go [hunk 3]
```diff
@@ -277,57 +576,96 @@ func getRemainingSleep(start time.Time, blockTime time.Duration, sleep time.Dura
...
-		}
-	} else if lastBatchHash == nil {
-		return nil, now, errors.New("lastBatch is not supposed to be nil")
-	} else {
-		if !bytes.Equal(c.lastBatchHash, lastBatchHash) {
-			return nil, now, errors.New("supplied lastBatch does not match with sequencer last batch")
-		}
+	c.lastBatchHashMutex.RLock()
+	lastBatchHash := c.lastBatchHash
+	c.lastBatchHashMutex.RUnlock()
```
Comment (MSevey) [inline sequencing/sequencer.go:599]: What's the rationale for not just holding the mutex for the entire operation? 
This method doesn't seem to be doing anything too long live, so just keeping it all atomic under one mutex lock feels reasonable and would avoid potential edge cases of the hashes being updated mid method call.

Comment (gupadhyaya) [reply sequencing/sequencer.go:599]: sure. no harm in keeping the lock.

**Review by MSevey — APPROVED** [review]

**Review by Manav-Aggarwal — APPROVED** [review]
utAck

**Pull request merged**